refactor(scripts): check_urls.vim to Lua + Treesitter

yochem opened the pull request:
Problem:
`scripts/check_urls.vim` manually matches urls in the help pages and then synchronously checks them via curl/wget/powershell. This is extremely slow (~5 minutes for Nvims runtime on my machine) and prone to errors in how the urls are matched.

Solution:
Incorporate the check for unreachable URLs in the validator of `src/gen/gen_help_html.lua`. Run it using:

```
NVIM_REQUEST_URLS=1 make lintdoc
```

justinmk commented:
> This is extremely slow (~5 minutes for Nvims runtime on my machine) and prone to errors in how the urls are matched.

Spawn it in a child job. I don't think vim.async is needed here, this script is not intended for interactive use, it's a build-time step.

justinmk commented on scripts/check_urls.lua:
```diff
@@ -0,0 +1,93 @@
+
+local query = ts.query.parse('vimdoc', '(url) @url')
+
+local function curl(url, cb)
+  local cmd = {
+    'curl',
+    '--silent',
```
can `vim.net.request` be used or what is missing from it?

yochem replied:
the redirect `-L` flag and that the error code value is a string, but I think that checking `err == nil` might be fine too.

justinmk commented on scripts/check_urls.lua:
```diff
@@ -0,0 +1,93 @@
+  }
+  vim.system(cmd, cb)
+end
+local request = async.wrap(2, curl)
```
`vim.system` is already async so don't need `vim.async`

yochem replied:
I'll rewrite it with vim.async removed and will report back

yochem commented:
> Spawn it in a child job. I don't think vim.async is needed here, this script is not intended for interactive use, it's a build-time step.

I added `vim.async` so a check for one url doesn't have to wait for the others. Is this the wrong use-case of `vim.async` and not supported by `vim.async`/coroutines? If it's not, then it's indeed not necessary.

yochem commented on runtime/lua/vim/net.lua:
```diff
@@ -25,7 +25,7 @@ function M.request(url, opts, on_response)
 
   -- Build curl command
-  local args = { 'curl' }
+  local args = { 'curl', '--max-time', '5' }
```
Not the final solution, but need to find a way to have this behaviour. Either by allowing curl arguments to `vim.net.request` or some callback-wizardry.

yochem replied:
#36164 Adds a `:cancel()` method, thus this should not be necessary anymore (the caller can handle the timeout). Until that is merged, this PR just prints a warning and doesn't cancel the tasks.

yochem committed:
refactor(scripts): port check_urls to Lua

Problem:
scripts/check_urls.vim manually matches urls in the help pages and then
synchronously checks them via curl/wget/powershell. This is extremely
slow (~5 minutes for Nvims runtime on my machine) and prone to errors in
how the urls are matched.

Solution:
Use Tree-sitter to find the urls in the help pages and `vim.net.request` to
check the responses. With a timeout of 20s per file and a max of 30
requests at the same time, this takes around 1 minute.

justinmk commented:
> Currently, the scripts is not used anywhere. Should it be called in `scripts/lintdoc.lua`?

Sure, I guess try it out in this PR since it needs conflict resolution anyway :)

I am expecting it to be somewhat flaky since it will be making lots of network requests... need to think about how to deal with that. Hopefully a network failure can be distinguished from a URL failure.

yochem committed:
refactor(scripts): check for unreachable URLs with lintdocurls

Problem:
check_urls.vim was slow and manually parsed the help files to find URLs.
The `make lintdoc` step only validates URLs by checking if they're not
http.

Solution:
Request URLs and check their response by running:

	make lintdocurls

Run this at 22:22 on every Friday. (Re)open an issue if there are any
unreachable URLs.

yochem commented on .github/workflows/doc_unreachable_urls.yml:
```diff
@@ -0,0 +1,37 @@
+          2>&1 NVIM_REQUEST_URLS=1 make lintdoc | sed -n '/invalid URLs/,/^}$/p' > $OUT_FILE
+          if [ -s $OUT_FILE ]; then
+            gh issue create \
+              --title "Unreachable URLs in help files" \
```
Changed it to:

```
gh issue reopen 36597
gh issue edit 36597 --body-file $OUT_FILE
```

I've created #36597 to serve as the issue that is reopened.

justinmk commented:
Cool! Thanks. Let's see what happens :)

yochem commented:
Yeah let's see this Friday. Of course, feel free to ping me anytime if issues may occur with the new CI action (or anything else related).

justinmk commented:
Anything we can do to improve  https://github.com/neovim/neovim/actions/runs/19774967693/job/56665991723 ? It failed as follows:


    sed: can't read 1i```: No such file or directory


Not sure what circumstances cause that, but perhaps the script can handle it better.
